compiler/rustc_lint/src/types.rs

```diff
@@ -921,11 +921,20 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         }
 
         match *ty.kind() {
-            ty::Adt(def, _) if def.is_box() && matches!(self.mode, CItemKind::Definition) => {
-                FfiSafe
-            }
-
             ty::Adt(def, substs) => {
+                if def.is_box() && matches!(self.mode, CItemKind::Definition) {
+                    if self.cx.layout_of(ty).unwrap().layout.size.bits()
+                        == self.cx.sess().target.pointer_width.into()
```

Comment:
I think `if !ty.boxed_ty().is_sized(tcx.at(DUMMY_SP), self.cx.param_env)` is a more correct way of checking if the Box is sized or not, that way you don't rely on the fact that DST>pointer, maybe in the future we might have DST<= pointer?

(also, I used `DUMMY_SP` because I couldn't find where the current span is in this function, but also I don't think the span matters much here?)

BUT take this with a grain of salt as I'm not a frequent contributor and not part of any team

Reply:
I think checking the layout size will cover more cases like boxed dyn trait? But we could check if the boxed type is Sized additionally to give more targeted error message I guess.

Reply:
Actually the boxed type in case of a boxed dyn trait is also unsized so your proposal would also cover it.
But I still think that for future proofing, having a check on the type layout size is also valuable to have.

Reply:
I agree with elichai, looks like `is_sized` is the thing that is supposed to be used here.

I'm not sure how the current logic will behave on encountering anything generic, for example. Hit the `.unwrap()`?

Reply:
Changed to `is_sized` as requested.